src/charm.py

```diff
@@ -80,11 +86,68 @@ def _on_config_changed(self, _):
         else:
             self.unit.status = WaitingStatus("Waiting for Pebble in workload container")
 
+    def _patch_auth_resources(self) -> None:
+        """Patch the Kubernetes auth resources created by Juju.
 
-    def _on_install(self, _):
-        """Placeholder for install event"""
-        pass
+        Raises:
+            ApiError: if patching any of the auth resources fails.
+        """
+        self.unit.status = MaintenanceStatus("Patching auth configuration")
+        client = Client()
+        context = {"namespace": self._namespace, "app_name": self._name}
+
+        with open(Path(self._src_dir) / "auth_manifests.yaml") as f:
+            for obj in codecs.load_all_yaml(f, context=context):
+                try:
+                    client.patch(
+                        type(obj), obj.metadata.name, obj, patch_type=PatchType.MERGE
+                    )
+                except ApiError as e:
+                    logging.error(
+                        f"Patching {obj.metadata.name} failed: {e.status.reason}"
+                    )
+                    raise
+        logging.info("Auth resources successfully patched.")
+
+    def _create_crds(self) -> None:
+        """Creates training-jobs CRDs.
+
+        Raises:
+            ApiError: if creating any of the CRDs fails.
+        """
+        self.unit.status = MaintenanceStatus("Creating CRDs")
+        client = Client()
+
+        with open(Path(self._src_dir) / "crds_manifests.yaml") as f:
+            for obj in codecs.load_all_yaml(f):
+                try:
+                    client.create(obj)
+                    logging.info(f"{obj.metadata.name} CRD successfully created.")
+                except ApiError as e:
+                    if e.status.reason == "AlreadyExists":
+                        logging.info(
+                            f"{obj.metadata.name} CRD already present. It will be used by the operator."
+                        )
+                    else:
+                        logging.error(
```

Comment:
I think we can skip this logging statement if we use `traceback.format_exc()` as in my other comment. The traceback includes the description of the object, including its name and why it failed. As an example where I tried to set an invalid label, I got this:

```python
... rest of traceback ...
lightkube.core.exceptions.ApiError: CustomResourceDefinition.apiextensions.k8s.io "profiles.kubeflow.org" is invalid: [metadata.labels: Invalid value: "\x00": name part must consist of alphanumeric characters, '-', '_' or '.', and must start and end with an alphanumeric character (e.g. 'MyName',  or 'my.name',  or '123-abc', regex used for validation is '([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9]'), metadata.labels: Invalid value: "!": a valid label must be an empty string or consist of alphanumeric characters, '-', '_' or '.', and must start and end with an alphanumeric character (e.g. 'MyValue',  or 'my_value',  or '12345', regex used for validation is '(([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9])?'), spec.versions[0].schema.openAPIV3Schema: Required value: schemas are required, spec.versions[1].schema.openAPIV3Schema: Required value: schemas are required]
```

I think that's enough information that we don't have to log it here as well.